### PR TITLE
[maven-central] Maven Central badge based on repo1.maven.org rather than search.maven.org

### LINKED ISSUE #846: Maven Central badge should be based on repo1.maven.org rather than search.maven.org
According to [this](http://central.sonatype.org/pages/ossrh-guide.html#releasing-to-central):

> Upon release, your component will be published to Central: this typically occurs within 10 minutes, though updates to search can take up to two hours.

So, besides the delay, [if the indexing jobs for search.maven.org are paused](https://issues.sonatype.org/browse/OSSRH-27247), it may take quite some time to see the last artifact version even though it has been actually released on Maven Central.

### PR DESCRIPTION
**Pull request by fabriziocucci**
According to [this](http://central.sonatype.org/pages/ossrh-guide.html#releasing-to-central):

> Upon release, your component will be published to Central: this typically occurs within 10 minutes, though updates to search can take up to two hours.

So, besides the delay, [if the indexing jobs for search.maven.org are paused](https://issues.sonatype.org/browse/OSSRH-27247), it may take quite some time to see the last artifact version even though it has been actually released on Maven Central.

Fixes #846.

### CODE REVIEW

**Comment by paulmelnikow** [comment]
Hi! Thank you for tackling this!

Would you be willing to add tests using the project's [new testing framework](https://github.com/badges/shields/tree/master/service-tests)?

**Comment by fabriziocucci** [comment]
I will give it a try, I'm sure it will be instructive! :nerd_face:

**Comment by paulmelnikow** [comment]
Feedback would be really appreciated! Don't hesitate to reach out if you have questions.

**Comment by fabriziocucci** [comment]
First question @paulmelnikow: since the new testing framework has been introduced only after my pull request, how do you want me to proceed? Rebase and (force) push the branch associated to this pull request? Merge master into my feature breanch? Other option?

I'm asking because I know every team has its own "rules" when it comes to not messing up with the history! :sweat_smile:

**Comment by paulmelnikow** [comment]
Hehe, indeed, every project is different. My preference is that you merge master into your branch. With all the commits in place, the discussion is easier to join midstream and after the fact, and means reviewers can just read the new commits. Think I read this in ESLint's guidelines, and it matches my experience with code review. Then it's on the maintainer to summarize the commit message at squash time. Though, it's helpful if feel you update the top comment and/or title to explain the "why" behind the changes.

**Comment by fabriziocucci** [comment]
@paulmelnikow I may be wrong but this should be complete now.

Please find below a couple of things I've found:

1. Following the [code coverage section](https://github.com/badges/shields/tree/master/service-tests#code-coverage) I wasn't able to get the code coverage because of this:

```javascript
$ npm run coverage:test:services -- -- --only=maven-central

> gh-badges@1.3.0 coverage:test:services /Users/fabriziocucci/git/shields
> istanbul cover _mocha --delay service-tests/runner/cli.js --dir coverage/services "--" "--only=maven-central"

=============================================================================
Writing coverage object [/Users/fabriziocucci/git/shields/coverage/services/coverage.json]
Writing coverage reports at [/Users/fabriziocucci/git/shields/coverage/services]
=============================================================================

=============================== Coverage summary ===============================
Statements   : 60.47% ( 78/129 )
Branches     : 26.92% ( 7/26 )
Functions    : 50% ( 9/18 )
Lines        : 60% ( 72/120 )
================================================================================
internal/process/next_tick.js:142
      throw new TypeError('callback is not a function');
      ^

TypeError: callback is not a function
    at process.nextTick (internal/process/next_tick.js:142:13)
    at Object.<anonymous> (/Users/fabriziocucci/git/shields/service-tests/runner/cli.js:9:4713)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/fabriziocucci/git/shields/node_modules/istanbul/lib/hook.js:107:24)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (internal/module.js:20:19)
    at /Users/fabriziocucci/git/shields/node_modules/mocha/lib/mocha.js:222:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/fabriziocucci/git/shields/node_modules/mocha/lib/mocha.js:219:14)
    at Mocha.run (/Users/fabriziocucci/git/shields/node_modules/mocha/lib/mocha.js:487:10)
    at Object.<anonymous> (/Users/fabriziocucci/git/shields/node_modules/mocha/bin/_mocha:459:18)
    at Module._compile (module.js:556:32)
    at Object.Module._extensions..js (module.js:565:10)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/fabriziocucci/git/shields/node_modules/istanbul/lib/hook.js:109:37)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Function.Module.runMain (module.js:590:10)
    at runFn (/Users/fabriziocucci/git/shields/node_modules/istanbul/lib/command/common/run-with-cover.js:122:16)
    at /Users/fabriziocucci/git/shields/node_modules/istanbul/lib/command/common/run-with-cover.js:251:17
    at /Users/fabriziocucci/git/shields/node_modules/istanbul/lib/util/file-matcher.js:68:16
    at /Users/fabriziocucci/git/shields/node_modules/async/lib/async.js:52:16
    at /Users/fabriziocucci/git/shields/node_modules/async/lib/async.js:361:13
    at /Users/fabriziocucci/git/shields/node_modules/async/lib/async.js:52:16
    at done (/Users/fabriziocucci/git/shields/node_modules/async/lib/async.js:246:17)
    at /Users/fabriziocucci/git/shields/node_modules/async/lib/async.js:44:16
    at /Users/fabriziocucci/git/shields/node_modules/async/lib/async.js:358:17
    at LOOP (fs.js:1739:14)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)

npm ERR! Darwin 15.6.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "coverage:test:services" "--" "--" "--only=maven-central"
npm ERR! node v6.6.0
npm ERR! npm  v3.10.3
npm ERR! code ELIFECYCLE
npm ERR! gh-badges@1.3.0 coverage:test:services: `istanbul cover _mocha --delay service-tests/runner/cli.js --dir coverage/services "--" "--only=maven-central"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the gh-badges@1.3.0 coverage:test:services script 'istanbul cover _mocha --delay service-tests/runner/cli.js --dir coverage/services "--" "--only=maven-central"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the gh-badges package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     istanbul cover _mocha --delay service-tests/runner/cli.js --dir coverage/services "--" "--only=maven-central"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs gh-badges
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls gh-badges
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/fabriziocucci/git/shields/npm-debug.log
```

2. regardless by the previous error, after executing that command the `coverage` folder is created and any subsequent execution of `npm test` will fail because that folder is scanned by eslint (i.e. `eslint '**/*.js'`).

3. if we want to be really strict, the tests don't cover the different color that the badge has if the latest maven version is a snapshot dependency or has version zero, i.e.

```javascript
...
      if (version === '0' || /SNAPSHOT/.test(version)) {
        badgeData.colorscheme = 'orange';
      } else {
        badgeData.colorscheme = 'blue';
      }
...
```

I look forward to receiving your feedback.

**Comment by fabriziocucci** [comment]
This is off-topic here but I'll write it down just in case I forget! :smile: 

If this PR goes through, as potential customization of this badge, it _could_ be worth considering to allow the user to specify `latest` (either release or snapshot), `release` (only release) or `snapshot` (only snapshot) as parameter. Since this PR relies on the [maven-metadata.xml](http://maven.apache.org/ref/3.2.5/maven-repository-metadata/repository-metadata.html), it would be really easy to extract that information. The common use case I'm thinking is the possibility of having two badges, one for the latest (stable) release version and one for the latest (unstable) snapshot.

File: service-tests/maven-central.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+const Joi = require('joi');
+const ServiceTester = require('./runner/service-tester');
+
+const t = new ServiceTester({ id: 'maven-central', title: 'Maven Central' });
+module.exports = t;
+
+t.create('latest version')
+  .get('/v/com.github.fabriziocucci/yacl4j.json')
```
Comment (paulmelnikow) [inline service-tests/maven-central.js]: Could you add a link to this project on maven central to aid future debugging?

File: server.js [hunk 2]
```diff
@@ -4254,24 +4254,26 @@ cache(function(data, match, sendBadge, request) {
...
     if (err != null) {
       badgeData.text[1] = 'inaccessible';
       sendBadge(format, badgeData);
       return;
     }
     try {
-      var data = JSON.parse(buffer);
-      var version = data.response.docs[0].latestVersion;
+      var versionRegExp = /\<latest\>(.*)\<\/latest\>/g;
+      var version = versionRegExp.exec(buffer)[1];
```
Comment (paulmelnikow) [inline server.js]: Instead of using a regex, would it be better to parse xml using `xml2js`, as in the Mozilla addons badge?

**Comment by paulmelnikow** [comment]
Thanks for the detailed feedback, which is very helpful. These tests look great! Will leave you a couple comments and then respond to your feedback.

**Comment by paulmelnikow** [comment]
I opened #975 with a fix for the two issues you encountered. As a temporary workaround until that is merged, you can:
- Run `./node_modules/.bin/istanbul cover node_modules/.bin/_mocha  service-tests/runner/cli.js --dir coverage/services "--" "--delay" "--only=maven-central"`
- Add the `.eslintignore` from that PR

> if we want to be really strict, the tests don't cover the different color that the badge has if the latest maven version is a snapshot dependency or has version zero

It's a good point you make. Testing colors would be a nice improvement. It's important for some badges, less important for others. Seems like it would be good here, since the color depends on a custom test.

To facilitate this, we could add the colors to the json badges. That seems a fine addition to the API.

> If this PR goes through, as potential customization of this badge, it _could_ be worth considering to allow the user to specify `latest` (either release or snapshot), `release` (only release) or `snapshot` (only snapshot) as parameter. Since this PR relies on the [maven-metadata.xml](http://maven.apache.org/ref/3.2.5/maven-repository-metadata/repository-metadata.html), it would be really easy to extract that information. The common use case I'm thinking is the possibility of having two badges, one for the latest (stable) release version and one for the latest (unstable) snapshot.

That's a nice idea. 😄  It could be done with an optional parameter in the badge, as in #959.

File: service-tests/maven-central.js (hunk 1 above)
Comment (fabriziocucci) [inline service-tests/maven-central.js]: You mean adding a comment like this:
```javascript
...
  .get('/v/com.github.fabriziocucci/yacl4j.json') // http://repo1.maven.org/maven2/com/github/fabriziocucci/yacl4j/
...
```
?

File: server.js (hunk 2 above)
Comment (fabriziocucci) [inline server.js]: That was actually my first implementation but then the idea of parsing the entire XML document only for getting the version seemed a bit too heavy.

Anyway, I'll bring it back if you think it's best! :sunglasses:

**Comment by fabriziocucci** [comment]
Replaced regexp with xml2js as per comment and added test to cover a potential parsing error.

Waiting for reply to the last pending comment!

File: service-tests/maven-central.js (hunk 1 above)
Comment (paulmelnikow) [inline service-tests/maven-central.js]: Yes, exactly!

File: server.js (hunk 2 above)
Comment (paulmelnikow) [inline server.js]: I thought I'd seen it somewhere recently!

Yea, I think so. It'll be more reliable, maintainable, and [secure](http://stackoverflow.com/a/1732454/893113), and encourages other contributors to follow suit. If XML parsing becomes a resource hog we could look for a more efficient parser.

**Comment by paulmelnikow** [comment]
Tangentially related: something's wrong with the code that is supposed to run the badge tests in Travis.

```
$ if [ "$TRAVIS_EVENT_TYPE" == pull_request ]; then npm run test:services -- --pr; fi
> gh-badges@1.3.0 test:services /home/travis/build/badges/shields
> mocha --delay service-tests/runner/cli.js "--pr"
PR: badges/shields#957
Title: [maven-central] Maven Central badge based on repo1.maven.org rather than search.maven.org
No services found. Nothing to do.
```

`[maven-central]` seems fine as a service id. Maybe the regex needs to be modified.

**Comment by fabriziocucci** [comment]
Added comment with the link to the project on Maven Central used in tests for the `maven-central` badge and opened [pull request to fix the regxexp used for matching services in pull requests](https://github.com/badges/shields/pull/978).

**Comment by paulmelnikow** [comment]
Merged #978 and re-triggered the build, and the Maven Central tests ran and passed on Travis. :+1:

The CRAN tests ran too though, so something's not quite right, though that needn't hold this up.

**Pull request merged**